pkg/kubelet/kubelet_pods.go

```diff
@@ -2345,3 +2361,59 @@ func (kl *Kubelet) cleanupOrphanedPodCgroups(pcm cm.PodContainerManager, cgroupP
 		go pcm.Destroy(val)
 	}
 }
+
+func (kl *Kubelet) runtimeClassSupportsRecursiveReadOnlyMounts(pod *v1.Pod) bool {
+	var runtimeClassName string
+	if pod.Spec.RuntimeClassName != nil {
+		runtimeClassName = *pod.Spec.RuntimeClassName
+	}
+	rtClasses := kl.runtimeState.runtimeClasses()
+	return runtimeClassSupportsRecursiveReadOnlyMounts(runtimeClassName, rtClasses)
+}
+
+// runtimeClassSupportsRecursiveReadOnlyMounts checks whether the runtime class supports recursive read-only mounts.
+// The kubelet feature gate is not checked here.
+func runtimeClassSupportsRecursiveReadOnlyMounts(runtimeClassName string, runtimeClasses []v1.RuntimeClass) bool {
+	for _, f := range runtimeClasses {
+		if f.Name == runtimeClassName {
+			return f.Features != nil && f.Features.RecursiveReadOnlyMounts != nil && *f.Features.RecursiveReadOnlyMounts
+		}
+	}
+	klog.ErrorS(nil, "unknown runtime class", "runtimeClassName", runtimeClassName)
+	return false
+}
+
+// resolveRecursiveReadOnly resolves the recursive read-only mount mode.
+func resolveRecursiveReadOnly(m v1.VolumeMount, runtimeSupportsRRO, featureGateEnabled bool) (bool, error) {
+	if m.RecursiveReadOnly == nil {
+		return false, nil
+	}
+	switch rroMode := *m.RecursiveReadOnly; rroMode {
+	case v1.RecursiveReadOnlyDisabled:
+		return false, nil
+	case v1.RecursiveReadOnlyIfPossible:
+		propNone := true
+		if m.MountPropagation != nil {
+			propNone = *m.MountPropagation == v1.MountPropagationNone
+		}
+		return featureGateEnabled && runtimeSupportsRRO && m.ReadOnly && propNone, nil
```

Comment:
Should we error here if:
1. Feature gate is not set OR
2. ReadOnly is not set OR
3. Mount Prop is not none
? i.e. IfPossible should only be tied to runtime support.

Reply:
I expect "IfPossible" to make the best effort to avoid errors, and just fall back to "Disabled"

Reply:
Because the IfPossible mode was introduced to allow creating a manifest that keeps compatibility on every node.
It ruins the purpose of IfPossible if we make it fallable.